Replace all labels with interned labels

geieredgar opened the pull request:
# Objective

First of all, this PR took heavy inspiration from #7760 and #5715. It intends to also fix #5569, but with a slightly different approach.


This also fixes #9335 by reexporting `DynEq`.

## Solution

The advantage of this API is that we can intern a value without allocating for zero-sized-types and for enum variants that have no fields. This PR does this automatically in the `SystemSet` and `ScheduleLabel` derive macros for unit structs and fieldless enum variants. So this should cover many internal and external use cases of `SystemSet` and `ScheduleLabel`.  In these optimal use cases, no memory will be allocated.

- The interning returns a `Interned<dyn SystemSet>`, which is just a wrapper around a `&'static dyn SystemSet`.
- `Hash` and `Eq` are implemented in terms of the pointer value of the reference, similar to my first approach of anonymous system sets in #7676.
- Therefore, `Interned<T>` does not implement `Borrow<T>`, only `Deref`. 
- The debug output of `Interned<T>` is the same as the interned value.

Edit: 
- `AppLabel` is now also interned and the old `derive_label`/`define_label` macros were replaced with the new interning implementation.
- Anonymous set ids are reused for different `Schedule`s, reducing the amount of leaked memory.

### Pros
- `InternedSystemSet` and `InternedScheduleLabel` behave very similar to the current `BoxedSystemSet` and `BoxedScheduleLabel`, but can be copied without an allocation.
- Many use cases don't allocate at all.
- Very fast lookups and comparisons when using `InternedSystemSet` and `InternedScheduleLabel`.
- The `intern` module might be usable in other areas.
- `Interned{ScheduleLabel, SystemSet, AppLabel}` does implement `{ScheduleLabel, SystemSet, AppLabel}`, increasing ergonomics.

### Cons
- Implementors of `SystemSet` and `ScheduleLabel` still need to implement `Hash` and `Eq` (and `Clone`) for it to work.

## Changelog

### Added

- Added `intern` module to `bevy_utils`.
- Added reexports of `DynEq` to `bevy_ecs` and `bevy_app`.

### Changed

- Replaced `BoxedSystemSet` and `BoxedScheduleLabel` with `InternedSystemSet` and `InternedScheduleLabel`.
- Replaced `impl AsRef<dyn ScheduleLabel>` with `impl ScheduleLabel`.
- Replaced `AppLabelId` with `InternedAppLabel`.
- Changed `AppLabel` to use `Debug` for error messages.
- Changed `AppLabel` to use interning.
- Changed `define_label`/`derive_label` to use interning. 
- Replaced `define_boxed_label`/`derive_boxed_label` with `define_label`/`derive_label`.
- Changed anonymous set ids to be only unique inside a schedule, not globally. 
- Made interned label types implement their label trait. 

### Removed

- Removed `define_boxed_label` and `derive_boxed_label`. 

## Migration guide

- Replace `BoxedScheduleLabel` and `Box<dyn ScheduleLabel>` with `InternedScheduleLabel` or `Interned<dyn ScheduleLabel>`.
- Replace `BoxedSystemSet` and `Box<dyn SystemSet>` with `InternedSystemSet` or `Interned<dyn SystemSet>`.
- Replace `AppLabelId` with `InternedAppLabel` or `Interned<dyn AppLabel>`.
- Types manually implementing `ScheduleLabel`, `AppLabel` or `SystemSet` need to implement:
  - `dyn_hash` directly instead of implementing `DynHash`
  - `as_dyn_eq`
- Pass labels to `World::try_schedule_scope`, `World::schedule_scope`, `World::try_run_schedule`. `World::run_schedule`, `Schedules::remove`, `Schedules::remove_entry`, `Schedules::contains`, `Schedules::get` and `Schedules::get_mut` by value instead of by reference.

alice-i-cecile commented:
Neat! I've added Controversial to all three of these PRs, to make sure the ECS SMEs consciously choose between them.

JoJoJet commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -34,7 +39,37 @@ pub trait SystemSet: DynHash + Debug + Send + Sync + 'static {
     /// Creates a boxed clone of the label corresponding to this system set.
-    fn dyn_clone(&self) -> Box<dyn SystemSet>;
+    fn dyn_leak(&self) -> &'static dyn SystemSet;
+    fn dyn_intern(&self) -> Interned<dyn SystemSet>
+    where
+        Self: Sized,
```
Doesn't requiring `Self: Sized` contradict the `dyn_` part of this function's name?

geieredgar replied:
You are right, I have chosen this name because I didn't want an conflict between `Intern::intern` (which is implemented by `dyn SystemSet`) and this method. 

Since this was only a convenience method to be able to write `x.dyn_intern()` instead of `(&x as &dyn SystemSet).intern()`, I removed it.

JoJoJet reviewed (commented):
I've scanned over this quickly, and it definitely seems like a promising approach. Very similar to #7760 but potentially preferable.

Just for clarity, what are the differences between `SystemSet::dyn_intern`, `Intern::intern`, and `IntoInterned::into_interned`? Is it possible to simplify or combine these methods some way?

geieredgar commented:
@JoJoJet I've replaced `SystemSet::dyn_intern` `Intern::intern` and `IntoInterned::into_interned` with the standard `Into<Interned<T>>::into`.

The functions that took `impl AsRef<dyn ScheduleLabel>` as parameter now take `impl Into<InternedScheduleLabel>`, which is implemented by the derive-macro for `ScheduleLabel` for `T` and `&T`, so it almost works the same as before.

I've also introduced a new trait and method `StaticRef::static_ref` which can be used to provide a consistent static reference to a value equal to `self`. This is used to further optimize the interning process, since interning a value for which a static reference can be provided doesn't need to lock the internal hash set anymore.

JoJoJet commented on crates/bevy_utils/src/intern.rs:
````diff
@@ -0,0 +1,310 @@
+    }
+}
+
+/// A type that can provide static references to equal values.
+pub trait StaticRef {
+    /// Returns a static reference to a value equal to `self`, if possible.
+    /// This method is used by [`OptimizedInterner::intern`] to optimize the interning process.
+    ///
+    /// # Invariant
+    ///
+    /// The following invariant must be hold:
+    ///
+    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` if and only if `a == b`
+    ///
+    /// where `ptr_eq` is defined as :
+    /// ```
+    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
+    ///     match (x, y) {
+    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
+    ///         _ => false,
+    ///     }
+    /// }
+    /// ```
+    fn static_ref(&self) -> Option<&'static Self>;
+}
+
+/// An optimized [`Interner`] for  types that implement [`StaticRef`].
+pub struct OptimizedInterner<T: ?Sized + 'static>(Interner<T>);
````
This is a good optimization, but why is `OptimizedInterner` separate from `Interner`? I can't imagine why you'd want an interner that doesn't perform this optimization. Would it be possible to further simplify this by combining these types, and potentially also combining `StaticRef` with `Leak`?

geieredgar replied:
I actually started with `static_ref` being a part of `Leak`, but then I realized that the invariant needs to be hold over all implementations of `Leak` and is therefore a property of the interned type and not of the `Leak` implementation.
Therefore I am against combining `StaticRef` with `Leak`.

Regarding the separation of `OptimizedInterner` and `Interner`, my thought was that it might be useful to create an interner for types that do not implement `StaticRef`, which could be the case for foreign types. But this is more of a hypothetical scenario, so I am open to combining those. If we later want to have an Interner for `str` we could also implement `StaticRef` for `str` instead.

geieredgar replied:
I've merged the `OptimizedInterner` into `Interner`.

JoJoJet commented on crates/bevy_app/src/app.rs:
```diff
@@ -379,9 +379,10 @@ impl App {
         schedule: impl ScheduleLabel,
         systems: impl IntoSystemConfigs<M>,
     ) -> &mut Self {
+        let schedule: InternedScheduleLabel = (&schedule as &dyn ScheduleLabel).into();
```
```suggestion
        let schedule = InternedScheduleLabel::from(&schedule as &dyn ScheduleLabel);
```

Not super important, but I think this syntax is preferable. Same for all of the other places this is used.

geieredgar replied:
I agree. Changed all places this was used.

JoJoJet reviewed (commented):
Can you also switch `AppLabel` to use interned labels? It still uses pseudo-stringly typed labels, and I'd rather not leave that relic around.

JoJoJet commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -30,8 +33,55 @@ pub trait SystemSet: DynHash + Debug + Send + Sync + 'static {
         false
     }
+
     /// Creates a boxed clone of the label corresponding to this system set.
     fn dyn_clone(&self) -> Box<dyn SystemSet>;
```
This method should probably be removed.

geieredgar replied:
We cannot remove this, it is used for leaking when `dyn_static_ref` returns `None`.

JoJoJet commented on crates/bevy_utils/src/intern.rs:
````diff
@@ -0,0 +1,310 @@
+    ///
+    /// # Invariant
+    ///
+    /// The following invariant must be hold:
+    ///
+    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` if and only if `a == b`
+    ///
+    /// where `ptr_eq` is defined as :
+    /// ```
+    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
+    ///     match (x, y) {
+    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
+    ///         _ => false,
+    ///     }
+    /// }
+    /// ```
````
```suggestion
    /// The following invariant must be hold:
    ///
    /// `ptr_eq(a.static_ref(), b.static_ref()) == true` only if `a == b`
    ///
    /// where `ptr_eq` is defined as :
    /// ```
    /// fn ptr_eq<T>(x: Option<&'static T>, y: Option<&'static T>) -> bool {
    ///     match (x, y) {
    ///         (Some(x), Some(y)) => std::ptr::eq(x, y),
    ///         _ => false,
    ///     }
    /// }
    /// ```
```

I believe the "if and only if" choice of words is slightly incorrect here. If `static_ref` returns `None`, then `ptr_eq` well evaluate to `false`, but `a` can still be equal to `b`.

geieredgar replied:
You are right, but the invariants are a bit more complex:

Basically, if we return `None` we are always safe. In the cases we return `Some`, we not only need to make sure that the pointer is identical for all equal values but also that the pointer is different for unequal values. 
And we cannot mix `Some` and `None` for equal values.

I resolved this by splitting the invariant into two parts.

geieredgar commented:
@JoJoJet Regarding `AppLabel`, is it okay if I do the switch in a follow-up PR?

JoJoJet commented:
I guess that's fine, but is there a reason not to do it now? As far as I know, you'd just have to change a `define_label!` to `define_interned_label!`, and update some callsites.

geieredgar commented:
OK, after giving it a second thought, I've decided to the `AppLabel` switch inside this PR. 

I also changed `AnonymousSet` to not use globally unique ids, and use ids only unique inside the given `Schedule`.  This means that we are not increasing the leaked memory indefinitely when the user creates `Schedules` repeatedly. Instead the amount of memory leaked by `AnonymousSet`s depends now on the `Schedule` that contained the most anonymous sets.

JoJoJet commented:
Thank you for implementing my suggestions. I'll try to review this more closely today.

JoJoJet commented:
I finally got around to this. I have a couple more ideas for how the API can be refined. Rather than going back-and-forth, I opened a PR to your branch: https://github.com/geieredgar/bevy/pull/1.

As always, feel free to voice any concerns you have with my suggestions.

JoJoJet commented on crates/bevy_utils/src/intern.rs:
```diff
@@ -0,0 +1,291 @@
+///
+/// For details on interning, see [the module level docs](self).
+///
+/// # Implementation details
+///
+/// This is just a reference to a value with an `'static` lifetime.
+//
+// This implements [`Copy`], [`Clone`], [`PartialEq`], [`Eq`] and [`Hash`] regardles of what `T`
+// implements, because it only uses the pointer to the value and not the value itself.
+// Therefore it MUST NEVER implement [`Borrow`](`std::borrow::Borrow`).
```
```suggestion
/// # Comparisons
/// 
/// Interned values use reference equality, meaning they implement [`Eq`]
/// and [`Hash`] regardless of whether `T` implements these traits.
/// Two interned values will only compare equal if they were interned using
/// the same [`Interner`] instance.
// NOTE: This type must NEVER implement Borrow since it does not obey that trait's invariants.
///
/// ```
/// # use bevy_utils::intern::*;
/// #[derive(PartialEq, Eq, Hash)]
/// struct Value(i32);
/// impl Leak for Value {
///     // ...
/// # fn leak(&self) -> &'static Self { Box::leak(Box::new(Value(self.0))) }
/// }
/// let interner_1 = Interner::new();
/// let interner_2 = Interner::new();
/// // Even though both values are identical, their interned forms do not
/// // compare equal as they use different interner instances.
/// assert_ne!(interner_1.intern(&Value(42)), interner_2.intern(&Value(42)));
/// ```
```

I think this behavior is important to document and test for. Feel free to tweak the phrasing.

JoJoJet reviewed (approved):
I think this PR is in a good state now! I just have a minor documentation suggestion.

JoJoJet commented:
> Interned<dyn SystemSet> does and probably should not implement SystemSet, I think for the same reasons that Box<dyn SystemSet> doesn't, but I haven't looked up all the discussions about that. Generally, Interned(x) and x should not be treated as the same thing when it comes to hash values and comparisons.

This can be removed from the cons list, and added to the pros.

JoJoJet commented on crates/bevy_utils/src/intern.rs:
```diff
@@ -0,0 +1,306 @@
+///
+/// Interned values use reference equality, meaning they implement [`Eq`]
+/// and [`Hash`] regardless of whether `T` implements these traits.
+/// Two interned values will only compare equal if they were interned using
```
```suggestion
/// Two interned values are only guaranteed to compare equal if they were interned using
```

A slight correction to this phrasing I wrote. Type that implement `static_ref` will compare equal regardless of the interner used.

JoJoJet commented on crates/bevy_macro_utils/src/lib.rs:
```diff
@@ -184,7 +184,47 @@ pub fn derive_boxed_label(input: syn::DeriveInput, trait_path: &syn::Path) -> To
+                quote! {
+                    match self {
+                        #(#variants)*
+                        _ => quote! { std::option::Option::None }
```
Is it right for there to be a `quote!` inside of a `quote!` like this? Is this branch covered by any test cases?

geieredgar replied:
No, this wasn't right. I should have added tests for the derive macros sooner, sorry. Now all types of structs and enum variants are tested.

nicopap reviewed (approved):
I think this approach is sound. Pretty neat.

I think it's possible to extend this so that it doesn't require leaking and a global `HashSet` for stuff like `OnEnter(state)`. But as-is I think the PR works.

I see a 1% perf increase on the `1000_schedule` benchmark, but I wouldn't put too much weight on it.

The only blocker for me is the duplicate code in macros & doc comment.

To summarize the approach taken here:

This pretty much defines a `static` reference per value added as label.
   * `Leak::leak`, despite the name, most of the time won't leak anything, but instead return a reference to a constant value). When the value can't be known at compile time, `leak` does leak using `Box::leak`.
   * The comparison of labels is fast because it only checks reference equality (`Interned<dyn Foo>` is a `&'static dyn Foo` and the `eq` impl compares the reference). this is a single `u64` comparison vs a trait object method call in the previous implementation.
   * Interestingly, each label type has its own `Interned<T>` type, and you can access the underlying `T`.
   * This is how `Debug` is implemented, it's a fairly natural extension of the implementation.

JoJoJet commented on crates/bevy_ecs/macros/src/set.rs:
```diff
@@ -21,12 +25,65 @@ pub fn derive_set(input: syn::DeriveInput, trait_path: &syn::Path) -> TokenStrea
+                    }
+                })
+                .collect();
+            if use_fallback_variant {
+                quote! {
+                    match self {
+                        #(#variants)*
+                        _ => std::option::Option::None
+                    }
+                }
+            } else {
+                quote! {
+                    match self {
+                        #(#variants)*
+                    }
+                }
+            }
```
I'd remove `use_fallback_variant` flag altogether:

```rust
quote! {
    match self {
        #(#variatns)*
        #[allow(unreachable_patterns)]
        _ => None,
    }
}
```

geieredgar replied:
I've changed the code to use
```rust
let fallback_variant = if variants.len() < data.variants.len() {
    quote!(_ => ::std::option::Option::None,)
} else {
    quote!()
};
quote! {
    match self {
        #(#variants)*
        #fallback_variant
    }
}
```

geieredgar replied:
Ok, I've switched to @JoJoJet's suggestion. I didn't see the comment before posting my own and refreshing the page.

JoJoJet commented on crates/bevy_utils/src/intern.rs:
```diff
@@ -0,0 +1,306 @@
+/// The implementation ensures that two equal values return two equal [`Interned<T>`] values.
+///
+/// To use an [`Interner<T>`], `T` must implement [`Leak`], [`Hash`] and [`Eq`].
+pub struct Interner<T: ?Sized + 'static>(OnceLock<RwLock<HashSet<&'static T>>>);
```
That was essentially the approach used in #4957. It ended up being too restrictive, since it forbade `OnEnter()` schedule labels (see #5569).

geieredgar replied:
We need this to support e.g.  `OnEnter`, `OnExit` or `AnonymousSet`. Note that we only have one `Interner` for `dyn SystemSet` and not for every type `T` that implements `SystemSet`.

nicopap replied:
Yeah. I think it is out of scope for this PR. I understand better now why it's not trivial to accomplish.

geieredgar commented on crates/bevy_utils/src/label.rs:
```diff
@@ -95,124 +94,98 @@ macro_rules! define_boxed_label {
-            fn clone(&self) -> Self {
-                self.dyn_clone()
+            /// Returns an [`Interned`](bevy_utils::intern::Interned) value corresponding to `self`.
+            fn intern(&self) -> ::bevy_utils::intern::Interned<dyn $label_trait_name>
```
Fixed.

geieredgar commented on crates/bevy_macro_utils/src/lib.rs:
```diff
@@ -196,127 +236,13 @@ pub fn derive_boxed_label(input: syn::DeriveInput, trait_path: &syn::Path) -> To
-        impl #impl_generics #trait_path for #ident #ty_generics #where_clause {
-            fn as_str(&self) -> &'static str {
-                #as_str
+            fn dyn_static_ref(&self) -> std::option::Option<&'static dyn #trait_path> {
```
I prefixed (hopefully) all occurrences of `std` with `::`. Thanks for pointing this out.

geieredgar commented on crates/bevy_derive/src/lib.rs:
```diff
@@ -203,10 +203,10 @@ pub fn derive_enum_variant_meta(input: TokenStream) -> TokenStream {
 ///
 /// This works only for unit structs, or enums with only unit variants.
 /// You may force a struct or variant to behave as if it were fieldless with `#[app_label(ignore_fields)]`.
```
Yes, it was wrong. Fixed that.

geieredgar commented on crates/bevy_macro_utils/src/lib.rs:
```diff
@@ -184,7 +184,47 @@ pub fn derive_boxed_label(input: syn::DeriveInput, trait_path: &syn::Path) -> To
+                }
+            }
+        }
+        syn::Data::Union(_) => quote! { std::option::Option::None },
```
I've extracted the code into a `static_ref_impl` function inside `bevy_macro_utils`.

geieredgar commented on crates/bevy_utils/src/label.rs:
```diff
@@ -95,124 +94,98 @@ macro_rules! define_boxed_label {
+            ///
+            /// # Invariant
+            ///
+            /// The following invariants must be hold:
```
Fixed. Thank you for spotting those.

geieredgar commented on crates/bevy_utils/src/label.rs:
````diff
@@ -95,124 +94,98 @@ macro_rules! define_boxed_label {
+            /// ```
+            ///
+            /// # Provided implementation
+            ///
````
I also removed the headings in `SystemSet` and `Leak`.

geieredgar commented on crates/bevy_ecs/macros/src/set.rs:
```diff
@@ -21,12 +25,65 @@ pub fn derive_set(input: syn::DeriveInput, trait_path: &syn::Path) -> TokenStrea
+                }
+            }
+        }
+        syn::Data::Union(_) => quote! { std::option::Option::None },
```
I agree, the derive macros will now fail again for unions.

geieredgar commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -202,14 +312,173 @@ mod tests {
         assert!(world.resource::<Flag>().0);
     }
+
+    #[test]
+    fn test_derive_schedule_label() {
```
Done.

JoJoJet commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -202,14 +310,197 @@ mod tests {
     }
+
+    #[test]
+    fn test_derive_schedule_label() {
```
Can you add test cases for non-unit fieldless structs and enums?

```rust
struct WeirdUnit1 { }

struct WeirdUnit2();

enum WeirdUnitEnum {
   Variant1 {},
    Variant2()
}
```

geieredgar replied:
I've added test cases for empty struct and tuple variants, both standalone and as part of an enum. I had to adjust `static_ref_impl` for it to work and stumbled across some weird behavior that is only present in the current stable channel, preventing me from using `&Self::EmptyTuple()` and having to use `&EnumLabel::EmptyTuple()` instead.  

On the Rust playground, [this](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=c544fa6b762881a53a46c8aae4e702a6) compiles, but [this](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=39468824e70ec79869792d83900d95f6) doesn't. Both versions work on the current beta channel.

JoJoJet replied:
You should be able to just do `&Self {}`, which would work for unit structs, as well as tuple and named structs with no fields.

geieredgar replied:
Ah, you're right. That's very useful to know.

nicopap commented on crates/bevy_macro_utils/src/static_ref.rs:
```diff
@@ -0,0 +1,40 @@
+                    let span = variant.span();
+                    let variant_ident = variant.ident.clone();
+                    if variant.fields.is_empty() {
+                        Some(quote_spanned! { span => Self::#variant_ident{} => ::std::option::Option::Some(&Self::#variant_ident{}), })
```
Do we know for sure that

1. Those are unique per variant
2. Two different instances of `Foo::Bar.dyn_static_ref()` always return the same `&'static` reference? It's important since the whole point is to compare the reference instead of the type+value.

It seems like it, given we pass the tests, but we are relying on an implicit behavior.

geieredgar replied:
Regarding the second point, you are right, I was relying on implicit behavior. [The Rust Reference](https://doc.rust-lang.org/reference/items/constant-items.html#constant-items) states that 

> References to the same constant are not necessarily guaranteed to refer to the same memory address.

Therefore I changed the implementation to use [static items](https://doc.rust-lang.org/reference/items/static-items.html#static-items) instead, which solves this issue, because

> All references to the static refer to the same memory location.

Regarding the first point, I don't see a reason why two variants should share the same reference / memory address, since their representation in memory is different. Of course two different ZSTs can have the same memory address, but the cast to `&dyn` will yield different vtable pointers for different types, so they would not be considered equal.

geieredgar replied:
Also, because of the restrictions on [generics for static items](https://doc.rust-lang.org/reference/items/static-items.html#statics--generics), types with generic type or const parameters will not have the `static_ref` optimization and have to go through the `Interner`, but will still not cause an allocation if they are ZST.

JoJoJet commented:
Does this fix #9335?

geieredgar commented:
> Does this fix #9335?

It does now, but with the cost of having to reexport `DynEq`. The second approach mentioned [here](https://github.com/bevyengine/bevy/issues/9335#issuecomment-1662022284) does not work with making `Interned<dyn T>` implementing `T` without running into the same issues again that were fixed by #8436.

JoJoJet commented:
Can you add Fixes #9335 to the PR description?

JoJoJet commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -2,57 +2,52 @@ use std::any::TypeId;
+    SCHEDULE_LABEL_INTERNER
+);
+
+define_label!(
+    /// Types that identify logical groups of systems.
+    SystemSet,
+    SYSTEM_SET_INTERNER,
+    {
+        /// Returns `Some` if this system set is a [`SystemTypeSet`].
+        fn system_type(&self) -> Option<TypeId> {
+            None
+        }
```
```suggestion
define_label!(
    /// Types that identify logical groups of systems.
    SystemSet,
    SYSTEM_SET_INTERNER,
    extra_methods: {
        /// Returns `Some` if this system set is a [`SystemTypeSet`].
        fn system_type(&self) -> Option<TypeId> {
            None
        }
```

This suggestion won't compile as-is, but to make it more clear what's happening here, can you add names to these blocks like `extra_methods` and `extra_methods_impl`?

geieredgar replied:
Done. Also rebased the PR on main again.

alice-i-cecile commented:
@geieredgar I'm pushing to have this in 0.12: are you able to resolve conflicts for us?

nicopap reviewed (commented):
I've had the time to think about it. And I'm still ambivalent. I don't think the idea is sound. Rust does not guarantee that different static addresses of ZSTs are different. This hinges on ZSTs static references being different if I'm not mistaken

Could we add tests to check that two different interned ZSTs do indeed not compare equal?

Something like

```rust
assert_ne!(UnitLabel1.intern(), UnitLabel2.intern());
```

Diddykonga commented:
Adding these here:
https://www.reddit.com/r/rust/comments/zfi61k/comment/izc2guh/?utm_source=share&utm_medium=web2x&context=3
https://www.reddit.com/r/rust/comments/zfi61k/comment/izceo6f/?utm_source=share&utm_medium=web2x&context=3

geieredgar commented:
@nicopap After finding https://github.com/rust-lang/rust/issues/106447 I think you are correct that this approach was unsound, because two equal labels could yield two different pointers. Therefore I removed the `static_ref` optimization completely, so interning will always go through the internal HashMap, guaranteeing that equal labels always return equal pointers.  

I also agree that two different labels cannot yield the same fat pointer when leaking, because then `a.dyn_eq(b)` would return `true`. And currently `ptr::eq` does compare both the address and the metadata, so it should work. However, as shown in https://github.com/rust-lang/rust/issues/106447 this behavior could change in the future.

geieredgar commented:
I've replaced the use of `std::ptr::eq` with a custom equality comparison, which first checks for equal type ids and then for equal data pointers. Therefore, we don't rely on the vtable pointers anymore.

alice-i-cecile commented:
@geieredgar this is in the milestone for 0.12, slated for this Saturday. I'm going to do a full review now, but I'm hopeful we can get this in and ship it. Can your resolve merge conflicts?

alice-i-cecile reviewed (approved):
Really cool stuff! The changes in `schedule.rs` and `intern.rs` are the only non-trivial changes: focus your review efforts there.

The breaking changes are acceptable and well-documented, the motivation is sound, I basically understand how this works, and this is incredibly well-tested. I'm comfortable approving this!

geieredgar commented:
Conflicts are resolved.

cart reviewed (approved):
I'm happy with this approach!

cart commented on crates/bevy_utils/src/intern.rs:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    fn ref_eq(&self, other: &Self) -> bool {
+        self.len() == other.len() && self.as_ptr() == other.as_ptr()
```
Just pushed as I doubt this will be controversial (and in the interest of moving forward).

cart replied:
I'll leave this open in case anyone wants to discuss.
